src/DataProtection/DataProtection/src/KeyManagement/KeyRingBasedDataProtector.cs

```diff
@@ -14,8 +15,237 @@
 
 namespace Microsoft.AspNetCore.DataProtection.KeyManagement
 {
+#if NETCOREAPP
     internal unsafe sealed class KeyRingBasedDataProtector : IDataProtector, IPersistedDataProtector
     {
+        public bool TryProtect(Span<byte> output, ReadOnlySpan<byte> plaintext, out int bytesWritten)
+        {
+            if (plaintext == null)
+            {
+                throw new ArgumentNullException(nameof(plaintext));
+            }
+
+            try
+            {
+                // Perform the encryption operation using the current default encryptor.
+                var currentKeyRing = _keyRingProvider.GetCurrentKeyRing();
+                var defaultKeyId = currentKeyRing.DefaultKeyId;
+                var defaultEncryptorInstance = currentKeyRing.DefaultAuthenticatedEncryptor;
+                CryptoUtil.Assert(defaultEncryptorInstance != null, "defaultEncryptorInstance != null");
+
+                if (_logger.IsDebugLevelEnabled())
+                {
+                    _logger.PerformingProtectOperationToKeyWithPurposes(defaultKeyId, JoinPurposesForLog(Purposes));
+                }
+
+                // We'll need to apply the default key id to the template if it hasn't already been applied.
+                // If the default key id has been updated since the last call to Protect, also write back the updated template.
+                var aad = _aadTemplate.GetAadForKey(defaultKeyId, isProtecting: true);
+
+                if (defaultEncryptorInstance.IsSpanEnabled())
+                {
+                    return defaultEncryptorInstance.TryEncrypt(output, plaintext, aad, out bytesWritten);
+                }
+
+                // We allocate a 20-byte pre-buffer so that we can inject the magic header and key id into the return value.
+                var retVal = defaultEncryptorInstance.Encrypt(
+                    plaintext: new ArraySegment<byte>(plaintext.ToArray()),
+                    additionalAuthenticatedData: new ArraySegment<byte>(aad),
+                    preBufferSize: (uint)(sizeof(uint) + sizeof(Guid)),
+                    postBufferSize: 0);
+
+                CryptoUtil.Assert(retVal != null && retVal.Length >= sizeof(uint) + sizeof(Guid), "retVal != null && retVal.Length >= sizeof(uint) + sizeof(Guid)");
+
+                // Make sure the output is large enough to hold the data
+                if (output.Length < retVal.Length)
+                {
+                    bytesWritten = 0;
+                    return false;
+                }
+
+                // At this point: retVal := { 000..000 || encryptorSpecificProtectedPayload },
+                // where 000..000 is a placeholder for our magic header and key id.
+
+                // Write out the magic header and key id
+                BinaryPrimitives.WriteUInt32BigEndian(retVal, MAGIC_HEADER_V0);
+                fixed (byte* pbRetVal = retVal)
+                {
+                    Write32bitAlignedGuid(&pbRetVal[sizeof(uint)], defaultKeyId);
```

Comment:
Can this use https://docs.microsoft.com/en-us/dotnet/api/system.guid.trywritebytes?view=net-5.0#System_Guid_TryWriteBytes_System_Span_System_Byte__?